File: CADBiS_jradius/src/cadbis/PacketsTodayLimits.java
package cadbis;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import cadbis.bl.Packet;
import cadbis.db.PacketDAO;
import cadbis.utils.DateUtils;

public class PacketsTodayLimits {
	protected final Logger logger = LoggerFactory.getLogger(getClass());
	protected HashMap<Integer, Double> packetsCoefs = null;
	protected HashMap<Integer, Long> dayTrafficLimits = null;
	protected HashMap<Integer, Long> monthTrafficLimits = null;
	protected Long restDaysCount = 0L;
	protected Long usedMonthTraffic = 0L;
	protected Long maximumMonthTraffic = 0L;
	protected Long restMonthTraffic = 0L;
	private Long allowedDayTraffic = 0L;
	
	protected void recalcTrafficLimits()
	{
		packetsCoefs = new HashMap<Integer, Double>();
		dayTrafficLimits = new HashMap<Integer, Long>();
		monthTrafficLimits = new HashMap<Integer, Long>();
		PacketDAO dao = new PacketDAO();
		BigInteger tmpMaximumMonthTraffic =  (BigInteger)dao.getSingleValueByQuery("select value from `cadbis_config` where name='max_month_traffic'","value");
		usedMonthTraffic = dao.getMonthTraffic();
		restDaysCount = Long.valueOf((DateUtils.getDaysInMonth() - DateUtils.getDOM()))+1;		
		if(usedMonthTraffic == null)
			usedMonthTraffic= 0L;
		List<Packet> packets = dao.getPacketsWithStats();
		if(tmpMaximumMonthTraffic!=null){
			maximumMonthTraffic = tmpMaximumMonthTraffic.longValue();
			restMonthTraffic = maximumMonthTraffic.longValue() - usedMonthTraffic.longValue();
			allowedDayTraffic  = (restMonthTraffic) / restDaysCount;
			
			double SumOfRangs = 0; 
			for(int i = 0; i< packets.size(); ++i)
				SumOfRangs += (double)packets.get(i).getSummarRang();
			for(int i = 0; i< packets.size(); ++i)
				packetsCoefs.put(packets.get(i).getGid(), ((double)packets.get(i).getSummarRang())/SumOfRangs);
			
			for(int i = 0; i< packets.size(); ++i)
			{
				Long dayLimit = Math.round((double)allowedDayTraffic * packetsCoefs.get(packets.get(i).getGid()));
				Long restPacketMonthTraffic = dayLimit*restDaysCount;
				monthTrafficLimits.put(packets.get(i).getGid(),restPacketMonthTraffic);
				if((packets.get(i).getExceed_times()+1)*dayLimit<=restPacketMonthTraffic)
					dayLimit *= packets.get(i).getExceed_times()+1; 
				dayTrafficLimits.put(packets.get(i).getGid(),dayLimit);
			}
		
		logger.info("maximumMonthTraffic = " + maximumMonthTraffic.longValue()/1024/1024+" Mb");
		logger.info("usedMonthTraffic = " + usedMonthTraffic.longValue()/1024/1024+" Mb");
		logger.info("restDaysCount = " + restDaysCount);
		logger.info("restMonthTraffic="+ restMonthTraffic/1024/1024+" Mb");
		logger.info("allowedDayTraffic="+ allowedDayTraffic/1024/1024+" Mb");
		for(int i = 0; i< packets.size(); ++i)
			logger.info("packet "+packets.get(i).getGid()+" '"+packets.get(i).getPacket()+"', coef="+packetsCoefs.get(packets.get(i).getGid())+",daylimit=" + dayTrafficLimits.get(packets.get(i).getGid())/1024/1024+"Mb of "+monthTrafficLimits.get(packets.get(i).getGid())/1024/1024+"Mb rest");
		
		}		
	}
	
	public Long getPacketDayTrafficLimit(Integer gid)
	{
		if(dayTrafficLimits == null)
			recalcTrafficLimits();
		return dayTrafficLimits.get(gid);
	}

	public Long getRestDaysCount() {
		return restDaysCount;
	}

	public Long getUsedMonthTraffic() {
		return usedMonthTraffic;
	}

	public Long getRestMonthTraffic() {
		return restMonthTraffic;
	}

	public Long getAllowedDayTraffic() {
		return allowedDayTraffic;
	}	
}

File: trunk/CADBiS_website/test/SM/CADBiS/PacketsTodayLimits.php
<?php
class PacketsTodayLimits
{
	/**
	 * CBilling object
	 * @var CBilling
	 */
	protected $BILL;
	
	protected $packets = null;
	protected $packetsCoefs = null;
	protected $allowedDayTraffic = null;
	protected $dayTrafficLimits = null;
	protected $monthTrafficLimits = null;	
	
	/**
	 * 
	 * @param CBilling $BILL
	 */
	public function __construct($BILL)
	{
		$this->BILL = $BILL;
		$this->packets = $BILL->GetTarifs();
	}
	
	/**
	 * recalculates packets limits
	 *
	 * @param array $packets
	 */
	protected function recalcTrafficLimits()
	{
		$config = $this->BILL->GetCADBiSConfig();
		$maximumMonthTraffic = $config['max_month_traffic'];  
		$monthAccts = $this->BILL->GetMonthTotalAccts();
		$usedMonthTraffic = $monthAccts['traffic'];
		$restDaysCount = date("t")-date("j");	
		$restMonthTraffic = $maximumMonthTraffic - $usedMonthTraffic;
		$this->allowedDayTraffic = ($restMonthTraffic) / $restDaysCount;
			
		$SumOfRangs = 0; 
		for($i = 0; $i< count($this->packets); ++$i)
			$SumOfRangs += $this->packets[$i]['rang'] * $this->packets[$i]['simuluse_sum'];
		for($i = 0; $i< count($this->packets); ++$i)
			$this->packetsCoefs[$this->packets[$i]['gid']] = ((double)$this->packets[$i]['rang'] * (double)$this->packets[$i]['simuluse_sum'])/(double)$SumOfRangs;
				
		for($i = 0; $i< count($this->packets); ++$i)
		{
			$dayLimit = round($this->allowedDayTraffic * $this->packetsCoefs[$this->packets[$i]['gid']]);
			$restPacketMonthTraffic = $dayLimit*$restDaysCount;
			$this->monthTrafficLimits[$this->packets[$i]['gid']]= $restPacketMonthTraffic;
			if($this->packets[$i]['exceed_times']*$dayLimit<=$restPacketMonthTraffic)
				$dayLimit *= $this->packets[$i]['exceed_times']+1;
			$this->dayTrafficLimits[$this->packets[$i]['gid']] = $dayLimit;
		}
	}
	
	/**
	 * Returns allowed traffic for a day
	 *
	 */
	public function getAllowedDayTraffic()
	{
		if($this->allowedDayTraffic == null)
			$this->recalcTrafficLimits();		
		return $this->allowedDayTraffic;
	}
	
	/**
	 * get traffic limit for a given gid
	 *
	 * @param int $gid
	 * @return int
	 */
	public function getPacketDayTrafficLimit($gid)
	{
		if($this->dayTrafficLimits == null)
			$this->recalcTrafficLimits();
		return $this->dayTrafficLimits[$gid];
	}
};